airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/manifest_reference_resolver.py

```diff
@@ -96,91 +96,102 @@ class ManifestReferenceResolver:
 
     ref_tag = "$ref"
 
-    def preprocess_manifest(self, manifest: Mapping[str, Any], evaluated_mapping: Mapping[str, Any], path: Union[str, Tuple[str]]):
+    def __init__(self, manifest: Mapping[str, Any]):
```

Comment:
I can see why we're going with the approach to have the set and the manifest defined on the factory itself, but it steel feels a little weird for them to be inserted into the factory during instantiation. When I think of a factory I see it as generic and the preprocess method would take input and emit output.

Granted we need to reference the manifest as we traverse the manifest so its probably easier to define it on the factory class itself. 

The other thing that feels weird is that when the `set()` is defined at the start, we're very reliant on correctly popping off visited elements after traversal and that at the end of preprocess the visited set is cleared.

As part of the `preprocess_manifest()` either when starting or finishing a run, we could clear the set so we know on each attempt we start clean. And the same could be said for assigning `self.manifest =` in that method (Although we would probably still instantiate empty values during init?). I'm not married to this approach either, but I think its worth a discussion because it stuck out to me that we're coupling the manifest to the factory pretty tightly in the refactor.

Reply:
So if you grep for `ManifestReferenceResolver` we actually only instantiate it and do the preprocessing once - when `ManifestDeclarativeSource` is initialized, and before the component factory is invoked at all. I'm wondering if that changes the discomfort with initializing it with the manifest + `visited` set. From there on out, the fully resolved config is accessed via `ManifestDeclarativeSource` and no additional calls are made to `preprocess_manifest`. That said if it feels more natural we can certainly consider clearing out the `visited` set after processing, and go back to the original design where we just passed the manifest into `preprocess_manifest`.

Reply:
Right so it's not a hard blocker by any means since like you said it's only used once. It's just a little bit of an awkward design. I don't think this part of flow would change, but one of the risks would be if we ever moved to using the factory or resolver multiple times, if we don't pop every element off, we might end up with some unexpected results accidentally using a set w/ stale elements.

How much lift is it to pass the manifest to the resolver like we used to be doing? The part that I noticed in your new implementation was that we use the `self.manifest` field in `_lookup_reference_value()`. If the lift isn't too big to continue the pass in the manifest and start w/ a clean execution of the function every time then I think that feels more intuitive. That being said, if trying to accommodate this makes the function too confusing, I think it is fine to leave as you already had it.

Reply:
It's no trouble at all - the code was in that state in one iteration, but I'd modified it thinking that it would be a little cleaner this way since I was having to pass `manifest` around as an argument in so many places. But to protect future-us from issues if we use the resolver multiple times I'm happy to change it back!

Reply:
Pushed this change.